### PR TITLE
Configuration of DMS schema components CDF upload

### PR DESCRIPTION
**Pull request by nikokaoja**
## Added

- Configuration for which DMS schema components are to be uploaded to CDF
- Configuration for which DMS schema components are to be removed to CDF
- Configuration how to handle existing CDF schema components during upload

## Changed
- Renamed `UploadDMSDataModel` to `ExportDMSSchemaComponentsToCDF` step. **Note** this is a breaking change, but
  since we are on 0. version, we can do this.
- Renamed `DeleteDMSDataModel` to `DeleteDMSSchemaComponents` step. **Note** this is a breaking change, but
  since we are on 0. version, we can do this.
- Renamed `ExportDMSDataModel` to `ExportDMSSchemaComponentsToYAML` step. **Note** this is a breaking change, but
  since we are on 0. version, we can do this.
- Renamed `DataModel` class to `DMSSchemaComponents` to better reflect the content of the class. **Note** this is a breaking change, but
  since we are on 0. version, we can do this.
- Step that waits for human approval timeout set to 1 day

### CODE REVIEW

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 1]
```diff
@@ -443,47 +444,23 @@ def as_api_view_property(property_: Property, space: str) -> MappedPropertyApply
...
-        """Checks if the data model exists in CDF.
+    def find_existing_spaces(self, client: CogniteClient) -> set:
+        """Checks if the spaces exist in CDF.
 
         Args:
             client: Cognite client.
 
         Returns:
-             True if the data model exists, False otherwise.
+            True if the containers exist, False otherwise.
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:454]: This does not match the type hint.

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 2]
```diff
@@ -443,47 +444,23 @@ def as_api_view_property(property_: Property, space: str) -> MappedPropertyApply
...
-            )
-
-        self.create_space(client)
-        self.create_containers(client)
-        self.create_views(client)
-        self.create_data_model(client)
-
-    def find_existing_data_model(self, client: CogniteClient) -> DataModelId | None:
-        """Checks if the data model exists in CDF.
+    def find_existing_spaces(self, client: CogniteClient) -> set:
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:447]:
```suggestion
    def find_existing_spaces(self, client: CogniteClient) -> set[str]:
```

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 3]
```diff
@@ -443,47 +444,23 @@ def as_api_view_property(property_: Property, space: str) -> MappedPropertyApply
...
 
-            return cdf_data_model.as_id()
-        return None
+        existing_spaces = set()
+
+        for space in self.spaces:
+            if client.data_modeling.spaces.retrieve(space):
+                existing_spaces.add(space)
+
+        return existing_spaces
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:463]: Lets not do a request per space.
```suggestion
        return set(client.data_modeling.spaces.retrieve(list(self.spaces)).as_ids())
```

Comment (doctrino) [reply cognite/neat/rules/exporter/_rules2dms.py:463]: Note the `find_existing_containers` and `find_existing_views` can equally be made into a simple one liner as this function.

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 4]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+            return cdf_data_model.as_id()
+        return None
+
+    def to_cdf(
+        self,
+        client: CogniteClient,
+        components_to_create: set | None = None,
+        existing_component_handling: Literal["fail", "skip", "update"] = "fail",
+    ) -> None:
+        """Write the the data model to CDF.
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:521]:
```suggestion
        """Write the data model to CDF.
```

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 5]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+
+        Args:
+            client: Connected Cognite client.
+            components_to_create: Which components to create. Takes set
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:525]: So we are introducing two new concept to neat with this PR?
* Schema: View+container+data model
* Component: Which is an item in the schema?

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 6]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+            - `fail`: if any component of the data model (DMS schema) already exists
+            - `skip`: skip DMS components that exist
+            - `update`: create DMS components that do not exist and update those that do exist
+
+            `update` policy is currently not implemented !
+
+        """
+
+        if components_to_create is None:
+            components_to_create = set(["all"])
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:547]:
```suggestion
        components_to_create = components_to_create  or  {"all"}
```

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 7]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+        if "space" in components_to_create or "all" in components_to_create:
+            self.create_space(client, existing_spaces)
+        if "container" in components_to_create or "all" in components_to_create:
+            self.create_containers(client, existing_containers)
+        if "view" in components_to_create or "all" in components_to_create:
+            self.create_views(client, existing_views)
+        if "data model" in components_to_create or "all" in components_to_create:
+            self.create_data_model(client, existing_data_model)
+
+    def create_space(self, client: CogniteClient, existing_spaces: set | None = None):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:569]: Lets rewrite this function to do it in one request. Not one request per space.

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 8]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+                _ = client.data_modeling.spaces.apply(SpaceApply(space=space))
+            else:
+                warnings.warn(
+                    f"Space {space} already exists in CDF! Skipping !!!",
+                    stacklevel=2,
+                )
+                logging.info(f"Space {space} already exists in CDF! Skipping !!!")
 
-    def create_containers(self, client: CogniteClient):
+    def create_containers(self, client: CogniteClient, existing_containers: set | None = None):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:581]: Same here, rewrite for a single request. Note that containers can have dependencies between them, and creating them one at at time is both inefficient and risk 400 if you do it in the wrong order.

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 9]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+                _ = client.data_modeling.containers.apply(self.containers[container_id])
+            else:
+                warnings.warn(
+                    f"Container {container_id} already exists! Skipping !!!",
+                    stacklevel=2,
+                )
+                logging.info(f"Container {container_id} already exists! Skipping !!!")
 
-    def create_views(self, client: CogniteClient):
+    def create_views(self, client: CogniteClient, existing_views: set | None = None):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:593]: Same here, rewrite to one request. For views this is really important as they typically have dependencies between each other, and creating them one-at-a-time, will fail with 400 if they are out of order or have circular dependencies. All at once solves this up to 1000 views (after that you have to be more careful but that should be handled by the SDK and not neat)

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 10]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+        if not existing_data_model:
+            logging.info(f"Creating data model {self.external_id} version {self.version} in space {self.space}")
+            _ = client.data_modeling.data_models.apply(
+                DataModelApply(
+                    name=self.name,
+                    description=self.description,
+                    space=self.space,
+                    external_id=self.external_id,
+                    version=self.version,
+                    views=list(self.views.values()),
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:619]: Here you are recreating the views. Is that intentional? What about just referencing them?
```suggestion
                    views=[view.as_id() for view in self.views],
```

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 11]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+                )
             )
-        )
+        else:
+            logging.info(f"Data model {self.space}:{self.external_id}/{self.version} already exists! Skipping !!!")
 
-    def remove_data_model(self, client: CogniteClient):
-        """Helper function to remove a data model, and all underlying views and containers from CDF.
+    def remove(self, client: CogniteClient, components_to_remove: set | None = None):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:625]: Type hint on return type. Is `mypy` skipped on this file?

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 12]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+        !!! note "Component Creation Policy"
+            Here is more information about the different component creation policies:
+            - `all`: all components of the data model will be created, meaning space, containers, views and data model
+            - `data model`: only the data model will be created
+            - `view`: only the views will be created
+            - `container`: only the containers will be created
         """
 
+        if components_to_remove is None:
+            components_to_remove = set(["all"])
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:641]:
```suggestion
        components_to_remove = components_to_remove  or {"all"}
```

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 13]
```diff
@@ -521,56 +498,186 @@ def find_existing_views(self, client: CogniteClient) -> set:
...
+        if "space" in components_to_remove or "all" in components_to_remove:
+            self.remove_spaces(client)
+
+    def remove_data_model(self, client: CogniteClient):
         if client.data_modeling.data_models.retrieve((self.space, self.external_id, self.version)):
-            logging.info(f"Removing data model {self.external_id} version {self.version} from space {self.space}")
+            logging.info(f"Removing data model {self.space}:{self.external_id}/{self.version}")
             _ = client.data_modeling.data_models.delete((self.space, self.external_id, self.version))
 
+    def remove_views(self, client: CogniteClient):
```
Comment (doctrino) [inline cognite/neat/rules/exporter/_rules2dms.py:656]: For all of these function, avoid removing one view, container and space at at time. Rewrite to one request

**Review by doctrino — CHANGES_REQUESTED** [review]
Nice renamings :) 

The two most critical things are
* Avoid introduction of `complex_configs` in workflow steps.
* Not do one-by-one creation of views, containers, spaces. It is only inefficient, but fairly error prone due to dependencies.

File: cognite/neat/workflows/steps/step_model.py [hunk 14]
```diff
@@ -47,6 +48,9 @@ class Step(ABC):
     def __init__(self, data_store_path: Path | None = None):
         self.log: bool = False
         self.configs: dict[str, str] = {}
+        self.complex_configs: dict[
+            str, typing.Any
+        ] = {}  # complex configs are meant for more complex configurations. Value can be any type.
```
[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/workflows/steps/step_model.py:53]: I requested from @alivinco to make this option in UI. He quickly prototype. 
This was request by a client to have more control over what they want to produce/remove in CDF.

[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/workflows/steps/step_model.py:53]: This is temporal option.

File: cognite/neat/workflows/model.py [hunk 15]
```diff
@@ -91,6 +91,7 @@ class WorkflowStepDefinition(BaseModel):
     ui_config: UIConfig = UIConfig()
     params: dict[str, str] = Field(default_factory=dict)  # System parameters
     configs: dict[str, Any] = Field(default_factory=dict)  # Step configurations
+    complex_configs: dict[str, Any] = Field(default_factory=dict)  # Complex step configurations
```
[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/workflows/model.py:94]: See comment above

File: cognite/neat/workflows/steps/step_model.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (alivinco) [reply cognite/neat/workflows/steps/step_model.py:53]: The reasoning behind introducing **complex_configs** is related to that **configs** is meant for values of string type ( in other place it's defined as dict[str,str])  and is sufficient in most cases and is easy to use. **complex_configs** was introduced to cover complex configurations that are not that common.

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 16]
```diff
@@ -236,8 +237,8 @@ def containers_from_rules(rules: Rules) -> dict[str, ContainerApply]:
                 container_property_id: str = cast(str, property_.container_property)
                 container_id = property_.container.id
 
-                api_container = DataModel.as_api_container(property_.container)
-                api_container_property = DataModel.as_api_container_property(property_)
+                api_container = DMSSchemaComponents.as_api_container(property_.container)
+                api_container_property = DMSSchemaComponents.as_api_container_property(property_)
```
[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:241]: Done!

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 17]
```diff
@@ -350,7 +351,7 @@ def views_from_rules(rules: Rules) -> dict[str, ViewApply]:
             Dictionary of ViewApply instances.
         """
         views: dict[str, ViewApply] = {
-            f"{rules.metadata.space}:{class_.class_id}": DataModel.as_api_view(
+            f"{rules.metadata.space}:{class_.class_id}": DMSSchemaComponents.as_api_view(
```
[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:354]: done

File: cognite/neat/rules/exporter/_rules2dms.py [hunk 18]
```diff
@@ -359,7 +360,7 @@ def views_from_rules(rules: Rules) -> dict[str, ViewApply]:
 
         # Create views from property-class definitions
         for row, property_ in rules.properties.items():
-            view_property = DataModel.as_api_view_property(property_, rules.metadata.space)
+            view_property = DMSSchemaComponents.as_api_view_property(property_, rules.metadata.space)
```
[the remark this answers is not in the document]
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:363]: done

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 1 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py]: Fixed

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 3 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py]: Done

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 5 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:517]: My proposal:

Schema contains following components:
- Spaces
- Containers
- Views
- Data Model

We had these from before, but it was not explicit, and all crammed under "data model".

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 6 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:539]: short and sweet, like it

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 7 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:561]: done!

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 8 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:573]: done

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 9 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:585]: great ! done implemented

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 10 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:611]: good catch ! need only ids!

Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:611]: fixed!

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 12 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:633]: implemented

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 13 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:648]: All at one implemented

File: cognite/neat/rules/exporter/_rules2dms.py (hunk 11 above)
Comment (nikokaoja) [reply cognite/neat/rules/exporter/_rules2dms.py:596]: Not skipped working,

**Pull request merged**